### ISSUE TITLE
Post-start script checking on cluster state

### ISSUE THREAD

**Issue by felixfriedrich** [bosh-elastic-stack/elasticsearch-boshrelease#8]
Hello, everyone!

tl;dr Why is it a good idea to check the cluster state in the post-start script of an individual node?

We ran into a situation in which is it problematic that the post-start script checks whether the cluster state is red or not.

I recently removed 2 data nodes from our ES cluster. This is apparently not a good idea, yet this mistake can be made easily, the current post-start script makes it impossible to recover from the situation. After having done the removal, a few shards vanished. This caused the cluster to be in a red state. The plan to recover from this was

1. Add a new data node
2. Attach a disk of one of the previously removed data nodes
3. See if the missing shards where on this disk; if not GOTO 1.

For step 2 to be successful the new data node needs to be stopped. `bosh stop` does that usually. In this case however it deleted the machine. After some paring on the situation we discovered that deleting the data node was only a side effect of the `bosh stop` command. The cause for this to happen is that step 1. was not successful due to the post-start script failing. The process on the machine however was running perfectly fine. Given this step failed bosh didn't update it's desired instance count. So when doing `bosh stop` bosh apparently compares the desired with the actual state and make sure both are equal. In this case that means that the last data node added would be removed again.

I would be interested in the reasons for the post-start script checking for the cluster state. Other than the situation in which we've been bitten by this and the ES process actually running perfectly fine, I can also think that a post-start script should only focus on the node in question, rather than the whole cluster. Because otherwise this would mean that the cluster has an influence on the question whether adding a new node was successful or not. In other words, I think a post-start script should only fail if there is something wrong with that particular node and only that particular node.

I am happy to be convinced otherwise and learn about the reasons for having added this check in the first place. In case my arguments are convincing, I am also happy to create a pull request with whatever improvement we agree on.

Thanks!

Felix

**Comment by felixfriedrich** [comment]
We now ran into another situation, which is related. We changed the stemcell of our deployment. So all data nodes have been re-created one by one. At a certain point the deployment failed. We found out that the following happened.

1. Node n was re-created. It held primary shard 1 for the log index of $TODAY.
2. Node n+1 was re-created. It help the replication shard 1 for log index of $TODAY.

Step 2 did not finish successfully as the post-start script checked for the cluster state to be red. Which was actually the case here as both copies of shard 1 were not present anymore in the cluster.
The reason is, we believe, that node n rejoined the cluster, but as the re-creation took longer than `index.unassigned.node_left.delayed_timeout`, the shards from node n have began to be re-created elsewhere before node n joined the cluster again. This process hasn't been finished before node n+1 was taken down. (At this point the post-start script should haven checked for that (is the cluster state yellow at this point?) and delayed the re-creation of node n+1 maybe?)

Now step 2 takes down node n+1. Here the post-start script check is performed and fails, yet it's already two late to serve a purpose, e.g. preventing data loss.

Loosing two shards at the same time is always unfortunate as this means lost data (given you have only two copies of each shard), in case it's $TODAY's index of logs it's even worse as Logstash fails to write into this index, which means even more lost data as new data cannot be stored. (Well Logstash buffers some logs, but still.)

Other than changing the post-start script in order to give node n time to fully join the cluster again, we identified these actions:

1. Raise `index.unassigned.node_left.delayed_timeout` such that the shards don't need to be created else where.
2. Use two replication shards, three in total in order to reduce the odds of this happening further. Note that this situation already only occurs if a primary shard and it's replication shards are located on nodes which happen to be re-created consecutively.

**Comment by making** [comment]
`post-start` basically comes from logsearch-boshrelease
https://github.com/cloudfoundry-community/logsearch-boshrelease/blob/develop/jobs/elasticsearch/templates/bin/post-start.erb

Pull Requests are welcome if you have suggestions to improve this behavior.

**Comment by making** [comment]
Does improving `post-start` as follows make sense?

* Wait until the cluster status gets `green`
* Make `index.unassigned.node_left.delayed_timeout` variable

**Comment by felixfriedrich** [comment]
At first glance, I think so. Not sure about all consequences this change might have, but I am sure we'll find out :-) Two thoughts though:

1. Changing `index.unassigned.node_left.delayed_timeout` prior to changes in nodes would be safer than permanently set the value. For example, increasing the disk size (scaling up) takes time as all data needs to be moved from the old disk to the new one. This would require `index.unassigned.node_left.delayed_timeout` to be several minutes. While this makes sense during this operation, it otherwise doesn't, or it at least more risky?
2. Is the `post-start` script the right place to wait for things? (Only checking once would definitely not work, right?) I am not sure about this actually - concept-wise. Is there a timeout in place after which this script is considered to be failed, if it hasn't terminated yet. Because if so, this might happen a lot. Re-balancing shards can take time depending on many factors (shard size, network limitations, physically as well as `indices.recovery.max_bytes_per_sec`.)

Thanks for making a suggestion :-) I am all in for improving this literately.

**Comment by making** [comment]
@felixfriedrich 
Just Released 0.12.0 that supports several new properties
https://github.com/bosh-elastic-stack/elasticsearch-boshrelease/releases/tag/0.12.0
Any feedback is welcome.

**Comment by antoniocascais** [comment]
Hi @making !

I'm a colleague of @felixfriedrich and we are using this release to deploy our production ES cluster.

We were having issues while deploying the cluster because it seemed like there was downtime (which, in theory, shouldn't) and also we were facing a possible red cluster in the following situation:

1) `node N` has shard `S1` and is taken out to apply changes
2) cluster is yellow
3) `node N` is up again but cluster didn't recognize that `node N` has shard `S1`
4) `node N+1` has shard replica `S1` and is taken out
5) cluster is red

The documentation on the internet says that we should disable shard allocation to perform a rolling upgrade.
We had a look into the drain script and it has
```
curl -s -X PUT -H 'Content-Type: application/json' http://127.0.0.1:9200/_all/_settings \
  -d '{"settings": {"index.unassigned.node_left.delayed_timeout": "<%= p("elasticsearch.recovery.delay_allocation_restart") %>"}}' \
  > /dev/null
```

This is not quite the same as disabling shard allocation (for example, it changes the settings on an index level not on a cluster level). In my personal opinion, disabling the relocation on a cluster level makes more sense and it's less error prone.

So what we did was to change the drain script api call from that to 
```
curl -s -X PUT -H 'Content-Type: application/json' http://127.0.0.1:9200/_cluster/settings \
  -d '{"persistent": {"cluster.routing.allocation.enable": "none"}}' \
	> /dev/null
```
And we also changed the post-start script api call to 
```
		curl -X PUT -s -H 'Content-Type: application/json' http://127.0.0.1:9200/_cluster/settings \
			-d '{"persistent": {"cluster.routing.allocation.enable": "all"}}'
    <% if p("elasticsearch.node.allow_data") and not p("elasticsearch.node.allow_master") %>
		if [ "${status}" = "200" ] && echo ${body} | grep '"status":"green"'; then
        rm ${out}
        # https://github.com/cloudfoundry-community/logsearch-boshrelease/pull/39
        exit 0
```

We need to enable shard relocation before checking for a green cluster, otherwise it will be stuck.

As far as we've seen so far, this does the trick:

1) Shard relocation is disabled
2) `node n` is out
3) `node n` is back up
4) shard relocation is enabled
5) we wait for a green cluster in the post start
6) cluster is green
7) `node n+1` starts to be deployed

What do you think about our changes? Shall we open a PR?

**Comment by antoniocascais** [comment]
Btw, the commit that I was talking about in the previous post is this one: https://github.com/springernature/elasticsearch-boshrelease/commit/99ee61fb4320f8212bae44d2785812459e2f4aa6

**Comment by making** [comment]
Yes, the PR that resolves your case is very welcome.
